### PR TITLE
Iterator pooling for snapshots

### PR DESCRIPTION
**Pull request by krishnangovindraj**
(Do not merge till we confirm the performance on TPCC loading. It currently looks 2x slower)

## Release notes: product changes
Introduces an iterator pool to reduce costs of creating & dropping rocked iterators.

### CODE REVIEW

File: storage/snapshot/snapshot.rs [hunk 1]
```diff
@@ -195,8 +198,9 @@ where
 }
 
 pub struct ReadSnapshot<D> {
-    storage: Arc<MVCCStorage<D>>,
     open_sequence_number: SequenceNumber,
+    iterator_pool: IteratorPool, // Must be declared & dropped before storage
+    storage: Arc<MVCCStorage<D>>,
```
Comment (krishnangovindraj) [inline storage/snapshot/snapshot.rs:203]: Alternately, the IteratorPool must hold a reference. This make the signature dependent on `<D>` which is ugly.

File: storage/snapshot/pool.rs
Comment (krishnangovindraj) [inline storage/snapshot/pool.rs:1]: Generic pool implementation

**Review by krishnangovindraj — COMMENTED** [review]
Don't merge yet. TPCC loading is 2x slower :(

**Comment by flyingsilverfin** [comment]
Closing in favor of https://github.com/typedb/typedb/pull/7281

**Pull request closed without merging**